File: smartadapter/src/main/java/smartadapter/widget/DefaultDiffUtilExtension.java
package smartadapter.widget;

/*
 * Created by Manne Öhlund on 2019-08-17.
 * Copyright (c) All rights reserved.
 */

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

import smartadapter.SmartRecyclerAdapter;

@SuppressWarnings("unchecked")
public class DefaultDiffUtilExtension extends DiffUtilExtension {

    private SmartRecyclerAdapter smartRecyclerAdapter;
    private DiffPredicate diffPredicate;
    private List oldList;
    private List newList;

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return diffPredicate.areItemsTheSame(oldList.get(oldItemPosition), newList.get(newItemPosition));
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return diffPredicate.areContentsTheSame(oldList.get(oldItemPosition), newList.get(newItemPosition));
    }

    @Override
    public void setSmartRecyclerAdapter(@NonNull SmartRecyclerAdapter smartRecyclerAdapter) {
        this.smartRecyclerAdapter = smartRecyclerAdapter;
    }

    @Override
    public void setDiffPredicate(@NonNull DiffPredicate diffPredicate) {
        this.diffPredicate = diffPredicate;
    }

    @Override
    public void diffSwapList(@NonNull List newList) {
        this.oldList = smartRecyclerAdapter.getItems();
        this.newList = newList;
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(this);
        diffResult.dispatchUpdatesTo(smartRecyclerAdapter);
        smartRecyclerAdapter.setItems(newList, false);
    }
}

File: smartadapter/src/main/kotlin/smartadapter/widget/DefaultDiffUtilExtension.kt
package smartadapter.widget

/*
 * Created by Manne Öhlund on 2019-08-17.
 * Copyright (c) All rights reserved.
 */

import androidx.recyclerview.widget.DiffUtil
import smartadapter.Position
import smartadapter.SmartRecyclerAdapter

@Suppress("UNCHECKED_CAST")
class DefaultDiffUtilExtension : DiffUtilExtension() {

    override var smartRecyclerAdapter: SmartRecyclerAdapter? = null
    private lateinit var diffPredicate: DiffPredicate<Any>
    private lateinit var oldList: List<Any>
    private lateinit var newList: List<Any>

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Position, newItemPosition: Position): Boolean {
        return diffPredicate.areItemsTheSame(
            oldList[oldItemPosition],
            newList[newItemPosition]
        )
    }

    override fun areContentsTheSame(oldItemPosition: Position, newItemPosition: Position): Boolean {
        return diffPredicate.areContentsTheSame(
            oldList[oldItemPosition],
            newList[newItemPosition]
        )
    }

    override fun setDiffPredicate(diffPredicate: DiffPredicate<*>) {
        this.diffPredicate = diffPredicate as DiffPredicate<Any>
    }

    override fun diffSwapList(newList: MutableList<*>) {
        smartRecyclerAdapter?.let { smartRecyclerAdapter ->
            this.oldList = smartRecyclerAdapter.getItems()
            this.newList = newList as MutableList<Any>
            val diffResult = DiffUtil.calculateDiff(this)
            diffResult.dispatchUpdatesTo(smartRecyclerAdapter)
            smartRecyclerAdapter.setItems(newList, false)
            smartRecyclerAdapter.updateItemCount()
        } ?: throw RuntimeException("SmartRecyclerAdapter is not set")
    }
}
